### DISCUSSION TITLE
Clone the Git-Hub Branch and Edit that

### DISCUSSION THREAD

**Discussion by GeoStork** [Jyers/Marlin#99] [Q&A] [answered]
Prior to finding this firmware, I was running a self compiled vanilla version of the latest Marlin release.  Followed the YouTube videos to get set up with VisualStudio, plugins, download the Marlin source, edit the configs so they work with the Ender 3 v2, etc.  Success.   Still very much a noob. Really new (really really new) to GitHub.   I'm interested to compile this fantastic firmware locally (to add a 5th filament type, insert default values specific to my set up, etc.).  Regarding doing this Jyers writes, "Of course instead of downloading the main marlin firmware **you will need to clone my GitHub branch** and edit that."    

What steps are involved to actually do this?  Is the desktop client for GitHub needed/desired?  Is cloning the same as creating a "fork"?    Not looking to shortcut or 'cheat' the learning process but would like to get familiar with and up to speed on the terms and the how to so I can keep learning.  Thanks in advance for sharing any tips and helpful guides.

**Comment by Schnabulation** [comment, marked as answer]
The beauty of GitHub is in the colaboration - meaning that it is really awesome to use when you are working with other people in projects. You can all work on the same code together. The GitHub desktop client (or commandline "git") is just the client that communicates with GitHub itself. It has a bunch of nice features like "merge" which (as the name implies) merges your code changes to the code already there.

But - you don't need to do this. You can just download the code using the *.zip download and you have everything you need. This is just a simple download of the code without any features of GitHub. If you want to make use of these you need to have the GitHub desktop client. Here are some important commands:

Clone --> downloads the code of a GitHub repository locally and connects them to the respective GitHub repo.
Pull --> When Jyers edits his code you can just run this command and it will update all files of your "cloned" repo locally.
Fork --> This one is special: if you want to have your own repository that is based on another repo, you can fork it. I personally have done that so I can have my own configuration files.
Merge --> When you have "forked" a repository and edited it, you can merge the original repository so all the changes Jyers did will get merged into your repo - without touching your edits / configs.

For now, you can just zip-download his firmware and adjust the configuration.h and configuration_adv.h files. If you want to learn GitHub I highly recommend you to fork the repo and play with it - it's so great! To get the latest fixes I can now just run "git merge upstream/Ender_3_v2_Extensible_UI" and THAT'S IT!

**Reply by GeoStork** [reply]
Perfect.   "Fork" is used if I want to edit the configuration files locally and "Merge" will apply repository changes without overwriting my config files?  What happens if a new feature is in the config files in the repository?  Is this to say that the config files (contents) are merged and repo changes applied with any changes/added lines on my end are preserved.  That makes sense if I understand correctly.  Thanks.

**Reply by Schnabulation** [reply]
Yes, exactly. This is the beauty of git: it will preserve your changes but add external changes in your files. So if a new feature needs code in the config files it will just add these lines but preserve your changes.

If you edited a line that the new feature needs to change there will be a merge conflict and you need to manually correct this.

Disclaimer: the names I used in my original reply are just the names of features so you can google the correct command. You can‘t just write „merge“ and that‘s it, the command needs more info.

**Reply by killlerpenguins** [reply]
@Schnabulation Hope you don't mind me jumping in here and asking as I am very new to all this and this is the exact part I am confused on as I have never used Git before. I have visual studio code installed as well as PlatformIO and I understand how to compile the firmware. I am just confused on how/what to exactly clone. Do I simply do this by selecting the <>Code tab, select the drop down for the green code button and download zip? (I provided a screenshot to show) Then simply unzip that file and access that folder in Visual Studio and edit what I need to change in the configuration.h folder then compile it? Is it that simple and I am just overthinking it or do I need to access a different folder location that I need to clone? I want to thank you in advance for taking the time to help if you can.

EDIT: After I posted this comment I was looking through more and noticed that there are other branches so then my next question is do I want to select the Ender_3_V2_Extensible_UI  (default) branch or do I want to select one of the other branches in the list? Screenshot posted below as well. Thank you again.

![image](https://user-images.githubusercontent.com/80071719/110039568-6c4e8a00-7d0f-11eb-929e-506638d99f80.png)

![image](https://user-images.githubusercontent.com/80071719/110043322-2d233780-7d15-11eb-85d9-d0795ef6ecb3.png)

**Reply by darco415** [reply]
Here some good info.....https://github.com/Jyers/Marlin/discussions/151 ....seems a little easier if you are use to using vscode

**Reply by killlerpenguins** [reply]
@darco415 Thanks for the reply. I'm actually not use to using VSCode as this is my first time using it but after watching a few videos on it I have a good understanding of the basic things. I had actually looked at that discussion you provided a few different times but being new to all this was really confusing so I had dismissed that. I would just use the precompiled files with BLTouch but issue is on my E3V2 I have the upgraded Z-Axis kit so I need to change my Z-Axis build volume height from the stock 250mm to 350mm.

**Reply by Schnabulation** [reply]
Hi there...

Basically you can just download the zip-file, as you mentioned. Then you have the most recent version and can use this to compile.

But, the beauty of git lies in its workflow: imagine Jyers changes code and you want to re-download the newest version. You would have to once again download the zip-file, extract it, adjust config-files and compile.
By using git (git bash or a GUI alternative) you can clone the repository locally (which just downloads all into a folder and connects it with GitHub) and make your changes. Now when Jyers updates his code, you can just do „git pull“ and it will download ONLY the changes. All your settings in your config-files will stay exactly the same.

Please go watch a git tutorial. There or not too many commands you need to know but it makes live so much easier...
